File: backend/app/schemas/agents.py
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field, field_validator

from app.schemas.llm import LLMUsageResponse


class AgentCatalogEntry(BaseModel):
    agent_key: str
    name: str
    category: str
    description: str
    status: str
    version: str
    capabilities: list[str]
    required_module: str
    orchestration_runtime: str = "langgraph"
    orchestration_features: list[str] = Field(default_factory=list)
    licensed: bool | None = None
    installed: bool | None = None
    enabled: bool | None = None
    license_gate: str = "unknown"


class AgentCatalogResponse(BaseModel):
    agents: list[AgentCatalogEntry]


class AgentInstanceCreateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    slug: str | None = Field(default=None, max_length=80)
    agent_key: str = Field(min_length=1, max_length=100)
    description: str | None = Field(default=None, max_length=500)
    visibility: str = Field(default="tenant", pattern="^(tenant|department|private)$")
    department_id: UUID | None = None
    owner_user_id: UUID | None = None
    model_routing_key: str | None = Field(default=None, max_length=120)
    model_key: str | None = Field(default=None, max_length=120)
    system_prompt: str | None = None
    config: dict[str, Any] = Field(default_factory=dict)
    metadata: dict[str, Any] = Field(default_factory=dict)

    @field_validator("slug")
    @classmethod
    def normalize_slug(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = value.strip().lower().replace(" ", "-")
        return normalized or None


class AgentInstanceUpdateRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=120)
    description: str | None = Field(default=None, max_length=500)
    status: str | None = Field(default=None, pattern="^(draft|active|disabled)$")
    visibility: str | None = Field(default=None, pattern="^(tenant|department|private)$")
    department_id: UUID | None = None
    owner_user_id: UUID | None = None
    model_routing_key: str | None = Field(default=None, max_length=120)
    model_key: str | None = Field(default=None, max_length=120)
    system_prompt: str | None = None
    config: dict[str, Any] | None = None
    metadata: dict[str, Any] | None = None


class AgentInstanceResponse(BaseModel):
    id: UUID
    tenant_id: UUID
    name: str
    slug: str
    agent_key: str
    module_key: str
    description: str | None
    status: str
    visibility: str
    department_id: UUID | None
    owner_user_id: UUID | None
    model_routing_key: str | None
    model_key: str | None
    system_prompt: str | None
    config: dict[str, Any]
    metadata: dict[str, Any]
    created_by: UUID | None
    created_at: Any
    updated_at: Any
    runnable: bool = True
    readiness: str = "ready"
    readiness_reasons: list[str] = Field(default_factory=list)
    model_available: bool = False
    knowledge_base_count: int = 0
    knowledge_enabled: bool = False


class AgentInstanceListResponse(BaseModel):
    agents: list[AgentInstanceResponse]


class WorkbenchAgentKnowledgeBaseSummary(BaseModel):
    id: UUID
    name: str
    description: str | None = None
    visibility: str
    status: str
    document_count: int = 0
    tags: list[str] = Field(default_factory=list)
    updated_at: Any


class WorkbenchAgentInstanceResponse(BaseModel):
    id: UUID
    name: str
    slug: str
    agent_key: str
    module_key: str
    description: str | None
    status: str
    visibility: str
    department_id: UUID | None
    category: str = "general"
    workflow_profile: str = "general"
    runnable: bool = True
    readiness: str = "ready"
    readiness_reasons: list[str] = Field(default_factory=list)
    model_profile: str | None = None
    model_policy: str
    model_available: bool = False
    knowledge_base_count: int = 0
    knowledge_enabled: bool = False
    knowledge_bases: list[WorkbenchAgentKnowledgeBaseSummary] = Field(default_factory=list)


class WorkbenchAgentInstanceListResponse(BaseModel):
    agents: list[WorkbenchAgentInstanceResponse]


class AgentGovernanceTargetItem(BaseModel):
    id: UUID
    label: str
    description: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class AgentGovernanceTargetsResponse(BaseModel):
    departments: list[AgentGovernanceTargetItem] = Field(default_factory=list)
    users: list[AgentGovernanceTargetItem] = Field(default_factory=list)
    knowledge_bases: list[AgentGovernanceTargetItem] = Field(default_factory=list)
    model_deployments: list[AgentGovernanceTargetItem] = Field(default_factory=list)


class AgentRunRequest(BaseModel):
    input: str = Field(min_length=1, max_length=12000)
    context: dict[str, Any] = Field(default_factory=dict)
    model_key: str | None = Field(default=None, max_length=120)
    routing_key: str | None = Field(default=None, max_length=120)
    max_tokens: int | None = Field(default=1024, ge=1, le=8192)
    mcp_server_keys: list[str] | None = Field(
        default=None,
        description="Optional MCP server_keys to mount for this run. None = use "
        "all active MCP servers in the tenant; empty list = disable MCP.",
    )


class AgentRunResponse(BaseModel):
    answer: str
    usage: LLMUsageResponse
    model_key: str
    request_id: str
    sources: list[dict[str, Any]] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(default_factory=dict)

File: frontend/src/lib/api/agents.ts
import { apiDelete, apiGet, apiPatch, apiPost } from "./core";
import type { LLMUsageResponse } from "./models";

export interface AgentCatalogEntryResponse {
  agent_key: string;
  name: string;
  category: string;
  description: string;
  status: string;
  version: string;
  capabilities: string[];
  required_module: string;
  orchestration_runtime: string;
  orchestration_features: string[];
  licensed: boolean | null;
  installed: boolean | null;
  enabled: boolean | null;
  license_gate: string;
}

export interface AgentCatalogResponse {
  agents: AgentCatalogEntryResponse[];
}

export interface AgentInstanceCreateRequest {
  name: string;
  slug?: string | null;
  agent_key: string;
  description?: string | null;
  visibility?: "tenant" | "department" | "private";
  department_id?: string | null;
  owner_user_id?: string | null;
  model_routing_key?: string | null;
  model_key?: string | null;
  system_prompt?: string | null;
  config?: Record<string, unknown>;
  metadata?: Record<string, unknown>;
}

export interface AgentInstanceUpdateRequest {
  name?: string | null;
  description?: string | null;
  status?: "draft" | "active" | "disabled" | null;
  visibility?: "tenant" | "department" | "private" | null;
  department_id?: string | null;
  owner_user_id?: string | null;
  model_routing_key?: string | null;
  model_key?: string | null;
  system_prompt?: string | null;
  config?: Record<string, unknown> | null;
  metadata?: Record<string, unknown> | null;
}

export interface AgentInstanceResponse {
  id: string;
  tenant_id: string;
  name: string;
  slug: string;
  agent_key: string;
  module_key: string;
  description: string | null;
  status: "draft" | "active" | "disabled";
  visibility: "tenant" | "department" | "private";
  department_id: string | null;
  owner_user_id: string | null;
  model_routing_key: string | null;
  model_key: string | null;
  system_prompt: string | null;
  config: Record<string, unknown>;
  metadata: Record<string, unknown>;
  created_by: string | null;
  created_at: string;
  updated_at: string;
  runnable?: boolean;
  readiness?: "ready" | "needs_configuration" | string;
  readiness_reasons?: string[];
  model_available?: boolean;
  knowledge_base_count?: number;
  knowledge_enabled?: boolean;
}

export interface AgentInstanceListResponse {
  agents: AgentInstanceResponse[];
}

export interface WorkbenchAgentInstanceResponse {
  id: string;
  name: string;
  slug: string;
  agent_key: string;
  module_key: string;
  description: string | null;
  status: "active";
  visibility: "tenant" | "department" | "private";
  department_id: string | null;
  category?: string;
  workflow_profile?: string;
  runnable: boolean;
  readiness: "ready" | "needs_configuration" | string;
  readiness_reasons: string[];
  model_profile: string | null;
  model_policy: "configured" | "system_default" | string;
  model_available: boolean;
  knowledge_base_count: number;
  knowledge_enabled: boolean;
  knowledge_bases: WorkbenchAgentKnowledgeBaseSummary[];
}

export interface WorkbenchAgentKnowledgeBaseSummary {
  id: string;
  name: string;
  description: string | null;
  visibility: "tenant" | "department" | "private" | string;
  status: "active" | "archived" | string;
  document_count: number;
  tags: string[];
  updated_at: string;
}

export interface WorkbenchAgentInstanceListResponse {
  agents: WorkbenchAgentInstanceResponse[];
}

export interface AgentGovernanceTargetItem {
  id: string;
  label: string;
  description: string | null;
  metadata: Record<string, unknown>;
}

export interface AgentGovernanceTargetsResponse {
  departments: AgentGovernanceTargetItem[];
  users: AgentGovernanceTargetItem[];
  knowledge_bases: AgentGovernanceTargetItem[];
  model_deployments: AgentGovernanceTargetItem[];
}

export interface AgentRunRequest {
  input: string;
  context?: Record<string, unknown>;
  model_key?: string | null;
  routing_key?: string | null;
  max_tokens?: number | null;
}

export interface AgentRunResponse {
  answer: string;
  usage: LLMUsageResponse;
  model_key: string;
  request_id: string;
  sources: Array<Record<string, unknown>>;
  metadata: Record<string, unknown>;
}

export const agentsApi = {
  getAgentGovernanceTargets: () => apiGet<AgentGovernanceTargetsResponse>("/api/v1/agents/governance-targets"),
  getAgentInstances: () => apiGet<AgentInstanceListResponse>("/api/v1/agents/instances"),
  getWorkbenchAgentInstances: () => apiGet<WorkbenchAgentInstanceListResponse>("/api/v1/agents/workbench/instances"),
  createAgentInstance: (payload: AgentInstanceCreateRequest) =>
    apiPost<AgentInstanceResponse, AgentInstanceCreateRequest>("/api/v1/agents/instances", payload),
  updateAgentInstance: (agentId: string, payload: AgentInstanceUpdateRequest) =>
    apiPatch<AgentInstanceResponse, AgentInstanceUpdateRequest>(`/api/v1/agents/instances/${agentId}`, payload),
  getAgentCatalog: () => apiGet<AgentCatalogResponse>("/api/v1/agents/catalog"),
  runAgent: (agentKey: string, payload: AgentRunRequest) =>
    apiPost<AgentRunResponse, AgentRunRequest>(`/api/v1/agents/${agentKey}/run`, payload),
};

// Agent Assignments

export interface AgentAssignment {
  id: string;
  agent_id: string;
  user_id: string;
  user_email: string;
  user_full_name: string | null;
  user_username: string | null;
  role: string;
  created_at: string;
  updated_at: string;
}

export interface AgentAssignmentListResponse {
  assignments: AgentAssignment[];
  total: number;
}

export interface AgentAssignmentBulkRequest {
  users: Array<{ user_id: string; role: string }>;
}

export interface MyAgent {
  id: string;
  agent_id: string;
  agent_name: string;
  agent_key: string;
  agent_slug: string;
  role: string;
  assigned_at: string;
}

export interface MyAgentListResponse {
  agents: MyAgent[];
  total: number;
}

// Agent Assignments
export async function listAgentAssignments(agentId: string): Promise<AgentAssignment[]> {
  const response = await apiGet<AgentAssignmentListResponse>(`/api/v1/agent-assignments/agents/${agentId}/users`);
  return response.assignments;
}

export async function assignUsersToAgent(
  agentId: string,
  users: Array<{ user_id: string; role: string }>,
): Promise<AgentAssignment[]> {
  const response = await apiPost<AgentAssignmentListResponse, AgentAssignmentBulkRequest>(
    `/api/v1/agent-assignments/agents/${agentId}/users`,
    { users },
  );
  return response.assignments;
}

export async function removeAgentUser(agentId: string, userId: string): Promise<void> {
  await apiDelete<void>(`/api/v1/agent-assignments/agents/${agentId}/users/${userId}`);
}

export async function listMyAgents(): Promise<MyAgent[]> {
  const response = await apiGet<MyAgentListResponse>("/api/v1/agent-assignments/my-agents");
  return response.agents;
}
